include/misc/math_extras_impl.h

```diff
@@ -0,0 +1,187 @@
+/*
+ * Copyright (c) 2019 Facebook.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/**
+ * @file
+ * @brief Inline implementation of functions declared in math_extras.h.
+ */
+
+#ifndef ZEPHYR_INCLUDE_MISC_MATH_EXTRAS_H_
+#error "please include <misc/math_extras.h> instead of this file"
+#endif
+
+#include <toolchain.h>
+
+/*
+ * Force the use of portable C code (no builtins) by defining
+ * PORTABLE_MISC_MATH_EXTRAS before including <misc/math_extras.h>.
+ * This is primarily for use by tests.
+ *
+ * We'll #undef use_builtin again at the end of the file.
+ */
+#ifdef PORTABLE_MISC_MATH_EXTRAS
+#define use_builtin(x) 0
+#else
+#define use_builtin(x) HAS_BUILTIN(x)
+#endif
+
+#if use_builtin(__builtin_add_overflow)
+static inline bool u32_add_overflow(u32_t a, u32_t b, u32_t *result)
+{
+	return __builtin_add_overflow(a, b, result);
+}
+
+static inline bool u64_add_overflow(u64_t a, u64_t b, u64_t *result)
+{
+	return __builtin_add_overflow(a, b, result);
+}
+
+static inline bool size_add_overflow(size_t a, size_t b, size_t *result)
+{
+	return __builtin_add_overflow(a, b, result);
+}
+#else /* !use_builtin(__builtin_add_overflow) */
+static inline bool u32_add_overflow(u32_t a, u32_t b, u32_t *result)
+{
+	u32_t c = a + b;
+
+	*result = c;
+
+	return c < a;
+}
+
+static inline bool u64_add_overflow(u64_t a, u64_t b, u64_t *result)
+{
+	u64_t c = a + b;
+
+	*result = c;
+
+	return c < a;
+}
+
+static inline bool size_add_overflow(size_t a, size_t b, size_t *result)
+{
+	size_t c = a + b;
+
+	*result = c;
+
+	return c < a;
+}
+#endif /* use_builtin(__builtin_add_overflow) */
+
+#if use_builtin(__builtin_mul_overflow)
+static inline bool u32_mul_overflow(u32_t a, u32_t b, u32_t *result)
+{
+	return __builtin_mul_overflow(a, b, result);
+}
+
+static inline bool u64_mul_overflow(u64_t a, u64_t b, u64_t *result)
+{
+	return __builtin_mul_overflow(a, b, result);
+}
+
+static inline bool size_mul_overflow(size_t a, size_t b, size_t *result)
+{
+	return __builtin_mul_overflow(a, b, result);
+}
+#else /* !use_builtin(__builtin_mul_overflow) */
+static inline bool u32_mul_overflow(u32_t a, u32_t b, u32_t *result)
+{
+	u32_t c = a * b;
+
+	*result = c;
+
+	return a > 0 && c / a != b;
+}
+
+static inline bool u64_mul_overflow(u64_t a, u64_t b, u64_t *result)
+{
+	u64_t c = a * b;
+
+	*result = c;
+
+	return a > 0 && c / a != b;
+}
+
+static inline bool size_mul_overflow(size_t a, size_t b, size_t *result)
+{
+	size_t c = a * b;
+
+	*result = c;
+
+	return a > 0 && c / a != b;
```

Comment:
It would be really helpfull if you could add () here and other returns above. Now it needs mental effort to read the line in order to figure out what it is doing.